### PR TITLE
cleanup `c/experimental/gradients/` part 4

### PR DESCRIPTION
**Pull request by vnghia**
@saxenasaurabh I have a lot of things that I would like to discuss with you, please take a look at it. Thank you ! 😄 

---

8f2852f: IMO, `mnist_grad_test` doesn't do anything really interesting. We have 
- 1 grad test for `MatMul` ( added in `math_grad_test`).
- 2 forward tests which test `ops` and aren't related to gradients.
- An another test for `ops` ( `ScalarMul` ).
- 1 `MNIST` training test but it only checks for the `status`.

So I removed it. I also removed `mnist_gradients_testutil` and `gradients_util` which are no longer used.

---

20db3b8: I am trying to add templates for `TestTensorHandleWithDims` and `TestScalarTensorHandle`. I think it won't break anything because these headers are used for testing only.

---

I am working on a bare minimum infrastructure for getting model trained fully in C++ ( optimizer, loss, metrics, ... ). Currently, I follow the interface of `model` in `keras`. You could see it here: https://github.com/vnvo2409/tensorflow/commit/9e42f7ea613c73b303829bfc833bc81d474d9218. With this infrastructure, `MNIST` achieves 96% accuracy ( with a real dataset downloaded from http://yann.lecun.com/exdb/mnist/ ) in `mnist_grad_test`. Hopefully we could use it for `resnet` soon. Below are the problems I have when building it.

---

How to pass `DataType` to `ops` ( e.g https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/cast#summary ). We might need it in the future. For an example, `MNIST` metric is essentially `tf.mean(tf.cast(tf.equal()))`

---

Should we rewrite the `ops` to something like 
```cpp
Status Equal(AbstractContext* ctx, AbstractTensorHandle* const x,
             AbstractTensorHandle* const y,
             absl::Span<AbstractTensorHandle*> outputs, const char* name,
             bool incompatible_shape_error = false);
```
and getting rid of this pattern
```cpp
Status Log1p(AbstractContext* ctx,
             absl::Span<AbstractTensorHandle* const> inputs,
             absl::Span<AbstractTensorHandle*> outputs, const char* name);
```

IMO, the former pattern is clearer and could help us avoid some confusion. An example, consider [`ops::Concat`](https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/concat#summary) ( because it both has `InputList` and `Input` ). 

---

When I used `AddInputList` for adding input to `ops::Concat`, it threw https://github.com/tensorflow/tensorflow/blob/c5012222b1d9accc07ea8fa4140c66039883ba2b/tensorflow/core/common_runtime/eager/eager_operation.cc#L403

`AddInputList` for `AddN` is working fine though.

---

[`ops::ApplyGradientDescent`](https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/apply-gradient-descent#summary) requires a `TF_FLOAT_REF` instead of `TF_FLOAT`. How to use it ?

---

Memory leaks: These code will leak because we lost access to the output of `MatMul` and `Relu` will replace it with another tensor without calling `Unref`. Is there an intuitive solution for this problem ? Currently, we could wrap it inside a `AbstractTensorHandlePtr` or save the pointer manually before passing it around but both ways require manual effort.

https://github.com/tensorflow/tensorflow/blob/c5012222b1d9accc07ea8fa4140c66039883ba2b/tensorflow/c/eager/mnist_gradients_testutil.cc#L120-L129

---

Thank you for your time once again ! 😄

### CODE REVIEW

**Comment by vnvo2409** [comment]
@kkimdev , @qqfish I see that you are the codeowers of `c_api_test_util`. Could you take a look at the template too ? Thank you ! 😄

**Comment by vnvo2409** [comment]
@saxenasaurabh Could you take a look please ? Thank you !

**Review by saxenasaurabh — COMMENTED** [review]
> @saxenasaurabh I have a lot of things that I would like to discuss with you, please take a look at it. Thank you ! 😄
> 
> [8f2852f](https://github.com/tensorflow/tensorflow/commit/8f2852fb95adbccfd6a936a8cd720faab17fb638): IMO, `mnist_grad_test` doesn't do anything really interesting. We have
> 
> * 1 grad test for `MatMul` ( added in `math_grad_test`).
> * 2 forward tests which test `ops` and aren't related to gradients.
> * An another test for `ops` ( `ScalarMul` ).
> * 1 `MNIST` training test but it only checks for the `status`.
> 
> So I removed it. I also removed `mnist_gradients_testutil` and `gradients_util` which are no longer used.

Sounds good. We have a MNIST MLP tested and benchmarked in `unified_api_test.py` so removing this should be fine.

> 
> [20db3b8](https://github.com/tensorflow/tensorflow/commit/20db3b8402d6713bb76bb33a27fa5c2b309943aa): I am trying to add templates for `TestTensorHandleWithDims` and `TestScalarTensorHandle`. I think it won't break anything because these headers are used for testing only.
> 
> I am working on a bare minimum infrastructure for getting model trained fully in C++ ( optimizer, loss, metrics, ... ). Currently, I follow the interface of `model` in `keras`. You could see it here: [vnvo2409@9e42f7e](https://github.com/vnvo2409/tensorflow/commit/9e42f7ea613c73b303829bfc833bc81d474d9218). With this infrastructure, `MNIST` achieves 96% accuracy ( with a real dataset downloaded from http://yann.lecun.com/exdb/mnist/ ) in `mnist_grad_test`. Hopefully we could use it for `resnet` soon. Below are the problems I have when building it.
> 

Looks very nice!

> How to pass `DataType` to `ops` ( e.g https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/cast#summary ). We might need it in the future. For an example, `MNIST` metric is essentially `tf.mean(tf.cast(tf.equal()))`
> 
> Should we rewrite the `ops` to something like
> 
> ```c++
> Status Equal(AbstractContext* ctx, AbstractTensorHandle* const x,
>              AbstractTensorHandle* const y,
>              absl::Span<AbstractTensorHandle*> outputs, const char* name,
>              bool incompatible_shape_error = false);
> ```
> 

I have not been careful about this but eventually we want to autogenerate these ops so we should have signatures that match the autogenerated ones. The autogenerated API will have a signature similar to python's `tf.raw_ops` API. We can get that for `Equal` by something like:

```
import inspect
import tensorflow as tf
inspect.getsource(tf.raw_ops.Equal).split("\n")[0]

> def equal(x, y, incompatible_shape_error=True, name=None):
```

So the API should be:
```
Status Equal(AbstractContext* ctx, AbstractTensorHandle* x,
             AbstractTensorHandle* y,
             absl::Span<AbstractTensorHandle*> outputs,
             bool incompatible_shape_error = true,
             const char* name = nullptr);
```


> and getting rid of this pattern
> 
> ```c++
> Status Log1p(AbstractContext* ctx,
>              absl::Span<AbstractTensorHandle* const> inputs,
>              absl::Span<AbstractTensorHandle*> outputs, const char* name);
> ```
> 
> IMO, the former pattern is clearer and could help us avoid some confusion. An example, consider [`ops::Concat`](https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/concat#summary) ( because it both has `InputList` and `Input` ).
> 
> When I used `AddInputList` for adding input to `ops::Concat`, it threw
> 
> https://github.com/tensorflow/tensorflow/blob/c5012222b1d9accc07ea8fa4140c66039883ba2b/tensorflow/core/common_runtime/eager/eager_operation.cc#L403

I will need to debug this. Let me know if this is blocking thing change. Else we can debug it with your other PR.

> 
> `AddInputList` for `AddN` is working fine though.
> 
> [`ops::ApplyGradientDescent`](https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/apply-gradient-descent#summary) requires a `TF_FLOAT_REF` instead of `TF_FLOAT`. How to use it ?

This op is no longer used for applying gradients. For Adam we have [ResourceApplyAdam](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/python/keras/optimizer_v2/adam.py;l=175;drc=c50af433c1be6bb235a7511b84337cfabf56323f). We could use this or maybe simply handwrite SGD for now `wt = wt - lr * grad`?

> 
> Memory leaks: These code will leak because we lost access to the output of `MatMul` and `Relu` will replace it with another tensor without calling `Unref`. Is there an intuitive solution for this problem ? Currently, we could wrap it inside a `AbstractTensorHandlePtr` or save the pointer manually before passing it around but both ways require manual effort.
> 
> https://github.com/tensorflow/tensorflow/blob/c5012222b1d9accc07ea8fa4140c66039883ba2b/tensorflow/c/eager/mnist_gradients_testutil.cc#L120-L129
> 

Yeah, we need to fix this. I haven't spent time thinking about a better API. Let me know if you have any ideas.

> Thank you for your time once again ! 😄

Thank you!

File: tensorflow/c/eager/c_api_test_util.h [hunk 1]
```diff
@@ -53,6 +56,27 @@ TFE_TensorHandle* TestTensorHandleWithDimsFloat(TFE_Context* ctx, float data[],
 TFE_TensorHandle* TestTensorHandleWithDimsInt(TFE_Context* ctx, int data[],
                                               int64_t dims[], int num_dims);
 
+// Return a tensor handle with given type, values and dimensions.
```
[the remark this answers is not in the document]
Comment (vnvo2409) [reply tensorflow/c/eager/c_api_test_util.h:59]: It is just a template so it has to be in the header file.

**Comment by vnvo2409** [comment]
> How to pass `DataType` to `ops` ( e.g https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/cast#summary ). We might need it in the future. For an example, MNIST metric is essentially `tf.mean(tf.cast(tf.equal()))`

How about this one ?

> When I used AddInputList for adding input to ops::Concat, it threw
https://github.com/tensorflow/tensorflow/blob/c5012222b1d9accc07ea8fa4140c66039883ba2b/tensorflow/core/common_runtime/eager/eager_operation.cc#L403
I will need to debug this. Let me know if this is blocking thing change. Else we can debug it with your other PR.

We can debug it in other PR.

I wonder if we should merge https://github.com/vnvo2409/tensorflow/commit/9e42f7ea613c73b303829bfc833bc81d474d9218 into `tensorflow` in another PR. WDYT ?

**Comment by vnvo2409** [comment]
@saxenasaurabh Could you take a look ? Thank you !

**Comment by saxenasaurabh** [comment]
> > How to pass `DataType` to `ops` ( e.g https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/cast#summary ). We might need it in the future. For an example, MNIST metric is essentially `tf.mean(tf.cast(tf.equal()))`
> 
> How about this one ?

Could we use [SetAttrType](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/eager/abstract_operation.h;l=93;drc=master)?

> 
> > When I used AddInputList for adding input to ops::Concat, it threw
> > https://github.com/tensorflow/tensorflow/blob/c5012222b1d9accc07ea8fa4140c66039883ba2b/tensorflow/core/common_runtime/eager/eager_operation.cc#L403
> > 
> > I will need to debug this. Let me know if this is blocking thing change. Else we can debug it with your other PR.
> 
> We can debug it in other PR.
> 
> I wonder if we should merge [vnvo2409@9e42f7e](https://github.com/vnvo2409/tensorflow/commit/9e42f7ea613c73b303829bfc833bc81d474d9218) into `tensorflow` in another PR. WDYT ?

Yeah, let's review that separately.

File: tensorflow/c/eager/c_api_test_util.h [hunk 2]
```diff
@@ -53,6 +56,27 @@ TFE_TensorHandle* TestTensorHandleWithDimsFloat(TFE_Context* ctx, float data[],
...
+  TFE_TensorHandle* th = TFE_NewTensorHandleFromTensor(ctx, t, status);
+  CHECK_EQ(TF_OK, TF_GetCode(status)) << TF_Message(status);
+  TF_DeleteTensor(t);
+  TF_DeleteStatus(status);
+  return th;
+}
+
+// Return a scalar tensor handle with given values.
+template <class T, TF_DataType datatype>
+TFE_TensorHandle* TestScalarTensorHandle(TFE_Context* ctx, const T value) {
```
Comment (saxenasaurabh) [inline tensorflow/c/eager/c_api_test_util.h:75]: Could we update other helpers in c_api_test_util.cc to use these new methods?

**Comment by vnvo2409** [comment]
> Could we use SetAttrType?

I haven't tried it yet but I think it won't work cause with `ops::Cast`, we have a separated attr ( https://www.tensorflow.org/api_docs/cc/struct/tensorflow/ops/cast/attrs#structtensorflow_1_1ops_1_1_cast_1_1_attrs ).

I think we need `AddInputType` ( and probably `TF_AddInputType` ). 

However, we don't need this ops now so we could leave it for later PR. I just want to bring it up so we could have some initial thoughts about it.

**Comment by saxenasaurabh** [comment]
> > Could we use SetAttrType?
> 
> I haven't tried it yet but I think it won't work cause with `ops::Cast`, we have a separated attr ( https://www.tensorflow.org/api_docs/cc/struct/tensorflow/ops/cast/attrs#structtensorflow_1_1ops_1_1_cast_1_1_attrs ).
> 
> I think we need `AddInputType` ( and probably `TF_AddInputType` ).
> 
> However, we don't need this ops now so we could leave it for later PR. I just want to bring it up so we could have some initial thoughts about it.

The destination type is an [attribute](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/ops/math_ops.cc;l=171;drc=86b694333ab8932e7394528e6bdf18ad15f44346) so I think we should be able to use `SetAttrType("DstT", DataType)` ?

**Comment by vnvo2409** [comment]
Ah I didn't see that. The function signature on the homepage did confuse me a lot. Unfortunately, my build cache is invalidated so I couldn't do anything until tomorrow.

File: tensorflow/c/eager/c_api_test_util.h (hunk 2 above)
Comment (vnvo2409) [reply tensorflow/c/eager/c_api_test_util.h:75]: Done

**Comment by vnvo2409** [comment]
@saxenasaurabh Sorry, you are right. We can use `SetAttrType` for `ops::Cast`. I have updated the PR as per your review. Please take a look ! Thank you !

File: tensorflow/c/eager/c_api_test_util.h (hunk 2 above)
Comment (saxenasaurabh) [reply tensorflow/c/eager/c_api_test_util.h:75]: Thanks for the cleanup. I was wondering if we can further use this to implement other functions in _this_ file e.g. https://github.com/tensorflow/tensorflow/blob/11ab65610238ab8469dd0a4c183ca4dc272ef61a/tensorflow/c/eager/c_api_test_util.cc#L108-L120

Comment (vnvo2409) [reply tensorflow/c/eager/c_api_test_util.h:75]: I did notice this file and other files in `c/eager`. But I am not sure if it will break something. @kkimdev What do you think ?

Comment (vnvo2409) [reply tensorflow/c/eager/c_api_test_util.h:75]: @saxenasaurabh , Since @kkimdev hasn't replied yet, could we merge this PR and I will send another PR if needed ?

Comment (saxenasaurabh) [reply tensorflow/c/eager/c_api_test_util.h:75]: I don't think this will break anything. Also these are test utilities so this is less risky of a refactor. Doing this as a follow up sounds good.

**Review by saxenasaurabh — APPROVED** [review]

File: tensorflow/c/eager/BUILD [hunk 3]
```diff
@@ -320,75 +320,6 @@ tf_cuda_cc_test(
...
-        "//tensorflow/c:tf_status_helper",
-        "//tensorflow/c/experimental/ops:array_ops",
-        "//tensorflow/c/experimental/ops:math_ops",
-        "//tensorflow/c/experimental/ops:nn_ops",
-        "//tensorflow/cc/profiler",
-        "//tensorflow/core:framework",
-        "//tensorflow/core:lib",
-        "//tensorflow/core:protos_all_cc",
-        "//tensorflow/core/lib/llvm_rtti",
-    ] + if_libtpu(
```
Comment (saxenasaurabh) [inline tensorflow/c/eager/BUILD:357]: Could you remove the `if_libtpu` import on line 6 above. The internal build complains otherwise.

Comment (vnvo2409) [reply tensorflow/c/eager/BUILD:357]: Done

**Review by saxenasaurabh — APPROVED** [review]
